src/Workspaces/Core/Portable/Classification/CommonClassificationService.cs

```diff
@@ -1,120 +1,71 @@
+        protected CommonClassificationService(Workspace workspace)
+            : base(workspace)
         {
         }
 
-        public override ImmutableArray<ClassifiedSpan> GetLexicalClassifications(SourceText text, TextSpan span, CancellationToken cancellationToken)
-        {
-            var result = ImmutableArray<ClassifiedSpan>.Empty;
-            var classifications = SharedPools.Default<List<ClassifiedSpan>>().Allocate();
-
-            try
-            {
-                this.AddLexicalClassifications(text, span, classifications, cancellationToken);
-                result = classifications.ToImmutableArray();
-            }
-            finally
-            {
-                SharedPools.Default<List<ClassifiedSpan>>().ClearAndFree(classifications);
-            }
-
-            return result;
-        }
-
-        protected abstract void AddLexicalClassifications(SourceText text, TextSpan textSpan, List<ClassifiedSpan> result, CancellationToken cancellationToken);
-
-        public override async Task<ImmutableArray<ClassifiedSpan>> GetSyntacticClassificationsAsync(Document document, TextSpan span, CancellationToken cancellationToken)
-        {
-            var workspace = document.Project.Solution.Workspace;
-            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
-            return GetSyntacticClassifications(tree, span, workspace, cancellationToken);
-        }
-
         public ImmutableArray<ClassifiedSpan> GetSyntacticClassifications(SyntaxTree tree, TextSpan span, Workspace workspace, CancellationToken cancellationToken)
         {
-            var syntaxClassifiers = this.GetDefaultSyntaxClassifiers();
-
-            var extensionManager = workspace.Services.GetService<IExtensionManager>();
-            var getNodeClassifiers = extensionManager.CreateNodeExtensionGetter(syntaxClassifiers, c => c.SyntaxNodeTypes);
-            var getTokenClassifiers = extensionManager.CreateTokenExtensionGetter(syntaxClassifiers, c => c.SyntaxTokenKinds);
-
-            var result = ImmutableArray<ClassifiedSpan>.Empty;
-            var classifications = SharedPools.Default<List<ClassifiedSpan>>().Allocate();
-
+            var spans = SharedPools.Default<List<ClassifiedSpan>>().Allocate();
+            var spanSet = SharedPools.Default<HashSet<ClassifiedSpan>>().Allocate();
             try
             {
-                AddSyntacticClassifications(tree, span, classifications, cancellationToken);
-                result = classifications.ToImmutableArray();
-            }
-            catch (Exception e) when (FatalError.ReportUnlessCanceled(e))
-            {
-                throw ExceptionUtilities.Unreachable;
+                var context = new ClassificationContext(spans, spanSet);
+                foreach (var provider in GetProviders())
+                {
+                    var commonProvider = provider as CommonClassificationProvider;
+                    if (commonProvider != null)
```

Comment:
A third party can implement a classification provider for C# or VB, but these would not be implemented as CommonClassificationProvider so they would get ignored on this path.

---
In reply to: [78664199](https://github.com/dotnet/roslyn/pull/13741#discussion_r78664199) [](ancestors = 78664199)